src/graph/sampler.cc

```diff
@@ -524,4 +525,198 @@ IdArray SamplerOp::RandomWalk(
+  actl_layer_sizes->push_back(node_mapping->size());
+  probabilities->insert(probabilities->end(), node_mapping->size(), 1);
+
+  size_t curr = 0;
+  size_t next = node_mapping->size();
+  unsigned int rand_seed = time(nullptr);
+  for (auto i = layer_sizes.rbegin(); i != layer_sizes.rend(); ++i) {
+    std::unordered_set<dgl_id_t> candidate_set;
+    for (auto j = curr; j != next; ++j) {
+      auto src = (*node_mapping)[j];
+      candidate_set.insert(indices + indptr[src], indices + indptr[src + 1]);
+    }
+
+    std::vector<dgl_id_t> candidate_vector;
+    std::copy(candidate_set.begin(), candidate_set.end(),
+              std::back_inserter(candidate_vector));
+
+    std::unordered_map<dgl_id_t, size_t> n_occurrences;
+    auto n_candidates = candidate_vector.size();
+    for (size_t j = 0; j != *i; ++j) {
+      auto dst = candidate_vector[rand_r(&rand_seed) % n_candidates];
+      if (!n_occurrences.insert(std::make_pair(dst, 1)).second) {
+        ++n_occurrences[dst];
+      }
+    }
+
+    for (auto const &pair : n_occurrences) {
+      node_mapping->push_back(pair.first);
+      float p = pair.second * n_candidates / static_cast<float>(*i);
+      probabilities->push_back(p);
+    }
+
+    actl_layer_sizes->push_back(node_mapping->size() - next);
+    curr = next;
+    next = node_mapping->size();
+  }
+  std::reverse(node_mapping->begin(), node_mapping->end());
+  std::reverse(actl_layer_sizes->begin(), actl_layer_sizes->end());
+  layer_offsets->push_back(0);
+  for (const auto &size : *actl_layer_sizes) {
+    layer_offsets->push_back(size + layer_offsets->back());
+  }
+}
+
+void ConstructFlows(const int64_t *indptr,
+                    const dgl_id_t *indices,
+                    const dgl_id_t *eids,
+                    const std::vector<dgl_id_t> &node_mapping,
+                    const std::vector<int64_t> &actl_layer_sizes,
+                    std::vector<int64_t> *sub_indptr,
+                    std::vector<dgl_id_t> *sub_indices,
+                    std::vector<dgl_id_t> *sub_eids,
+                    std::vector<dgl_id_t> *flow_offsets,
+                    std::vector<dgl_id_t> *edge_mapping) {
+  auto n_flows = actl_layer_sizes.size() - 1;
+  sub_indptr->insert(sub_indptr->end(), actl_layer_sizes.front() + 1, 0);
+  flow_offsets->push_back(0);
+  int64_t first = 0;
+  for (size_t i = 0; i < n_flows; ++i) {
+    auto src_size = actl_layer_sizes[i];
+    std::unordered_map<dgl_id_t, dgl_id_t> source_map;
+    for (int64_t j = 0; j < src_size; ++j) {
+      source_map.insert(std::make_pair(node_mapping[first + j], first + j));
+    }
+    auto dst_size = actl_layer_sizes[i + 1];
+    for (int64_t j = 0; j < dst_size; ++j) {
+      auto dst = node_mapping[first + src_size + j];
+      typedef std::pair<dgl_id_t, dgl_id_t> id_pair;
+      std::vector<id_pair> neighbor_indices;
+      for (int64_t k = indptr[dst]; k < indptr[dst + 1]; ++k) {
+        auto ret = source_map.find(indices[k]);
```

Comment:
There are a lot of hashtable lookups here. We might need to accelerate it in the future. Please add a comment here.